kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/KieServerImpl.java

```diff
@@ -918,7 +931,81 @@ public void addServerStatusMessage(KieServerInfo kieServerInfo) {
     public PolicyManager getPolicyManager() {
         return this.policyManager;
     }
+    
+    public boolean isKieServerReady() {
+        return kieServerReady.get();
+    }
+
+    public void markAsReady() {
+        kieServerReady.set(true);
+        logger.info("KieServer {} is ready to receive requests", KieServerEnvironment.getServerId());
+    }
+    
+    public List<Message> healthCheck(boolean report) throws IllegalStateException {
+        List<Message> healthMessages = new ArrayList<>();
+        long start = System.currentTimeMillis();
+        if (!isKieServerReady()) {
+            healthMessages.add(new Message(Severity.ERROR, String.format("KIE Server '%s' is not ready to serve requests",
+                                                                         KieServerEnvironment.getServerId())));
+        }
+        
+        if (report) {
+            List<String> mainInfo = new ArrayList<>();
+            mainInfo.add(String.format("KIE Server '%s' is ready to serve requests %s", 
+                                       KieServerEnvironment.getServerId(), 
+                                       isKieServerReady()));
+            mainInfo.add("Server is up for " + calculateUptime());
+            
+            Message header = new Message(Severity.INFO, mainInfo);
+            healthMessages.add(header);
+        }
+        // first check of KIE Server's containers if any of them is in failed state
+        for (KieContainerInstance container : getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.FAILED)) {
+                healthMessages.add(new Message(Severity.ERROR, String.format("KIE Container '%s' is in FAILED state",
+                                                                             container.getContainerId()) ));
+            }
+        }
+        // next check all extensions for their health
+        for (KieServerExtension extension : getServerExtensions()) {
+            List<Message> extensionMessages = extension.healthCheck(report);
+            healthMessages.addAll(extensionMessages);        
+        }
+        
+        if (report) {           
+            Message footer = new Message(Severity.INFO, "Health check done in " + (System.currentTimeMillis() - start) + " ms");
+            healthMessages.add(footer);
+        }
+        
+        return healthMessages;
+ 
+    }
+    
+    private String calculateUptime() {
```

Comment:
Did you consider using java.time.Duration for these calculations?

Reply:
I actually did and the problem is that it does not take into account already used parts of it, see example:

```
0 days, 0 hours, 30 minutes, 1800 seconds
0 days, 0 hours, 30 minutes, 0 seconds
```
first one is done with duration.toMinutes and duration.toMillis()/1000 as there seems to be missing toSeconds method.

Or did you have something else in mind?

Reply:
I see. Was thinking about using some formatter to convert Duration to some string representation, but it seems that there isn't any out of the box.
Current approach seems ok.

Reply:
You can use plain toString method on a duration, it will generate for example PT41M29.064S. It is ISO 8601 format. In fact, Duration's toString method does it manually with div and mod too.

Reply:
@MarianMacik correct, but it's way less readable in my opinion, especially for those that are not familiar with ISO date format.

Reply:
There are also easy to use things [DurationFormatUtils](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/time/DurationFormatUtils.html) in commons-lang3

Reply:
@jhrcek you rock! replaced with DurationFormatUtils, thanks!